core/pkg/store/store.go

```diff
@@ -131,146 +213,176 @@ func (f *Store) String() (string, error) {
 }
 
 // GetAll returns a copy of the store's state (copy in order to be concurrency safe)
-func (f *Store) GetAll(_ context.Context) (map[string]model.Flag, model.Metadata, error) {
-	txn := f.db.Txn(false)
-
+func (s *Store) GetAll(ctx context.Context, selector Selector) (map[string]model.Flag, model.Metadata, error) {
 	flags := make(map[string]model.Flag)
-	it, err := txn.Get("flags", "id")
-
-	if err != nil {
-		return flags, model.Metadata{}, err
+	queryMeta := model.Metadata{}
+	if selector != nil && !selector.IsEmpty() {
+		queryMeta = selector.SelectorToMetadata()
 	}
+	it, err := s.selectOrAll(selector)
 
-	for obj := it.Next(); obj != nil; obj = it.Next() {
-		flag := obj.(model.Flag)
-		flags[flag.Key] = flag
+	if err != nil {
+		s.logger.Error(fmt.Sprintf("flag query error: %v", err))
+		return flags, queryMeta, err
 	}
-
-	return flags, f.getMetadata(), nil
+	flags = s.collect(it)
+	return flags, queryMeta, nil
 }
 
 // Update the flag state with the provided flags.
-func (f *Store) Update(
+func (s *Store) Update(
 	source string,
-	selector string,
 	flags map[string]model.Flag,
 	metadata model.Metadata,
 ) (map[string]interface{}, bool) {
-	notifications := map[string]interface{}{}
 	resyncRequired := false
 
-	txn := f.db.Txn(true)
+	if source == "" {
+		panic("source cannot be empty")
+	}
+
+	priority := slices.Index(s.sources, source)
+	if priority == -1 {
+		// this is a hack to allow old constructors that didn't pass sources, remove when we remove "NewFlags" constructor
+		if !slices.Equal(s.sources, noValidatedSources) {
+			panic(fmt.Sprintf("source %s is not registered in the store", source))
+		}
+		// same as above - remove when we remove "NewFlags" constructor
+		priority = 0
+	}
+
+	txn := s.db.Txn(true)
 	defer txn.Abort()
-	storedFlags, _, _ := f.GetAll(context.Background())
-
-	f.mx.Lock()
-	f.setSourceMetadata(source, metadata)
-	for key, v := range storedFlags {
-		if v.Source == source && v.Selector == selector {
-			if _, ok := flags[key]; !ok {
-				// flag has been deleted
-				_, err := txn.DeleteAll("flags", "id", key)
-				if err != nil {
-					f.logger.Error(fmt.Sprintf("error deleting flag: %s, %v", key, err))
-					continue
-				}
 
-				notifications[key] = map[string]interface{}{
-					"type":   string(model.NotificationDelete),
-					"source": source,
-				}
-				resyncRequired = true
-				f.logger.Debug(
-					fmt.Sprintf(
-						"store resync triggered: flag %s has been deleted from source %s",
-						key, source,
-					),
-				)
-				continue
+	// get all flags for the source we are updating
+	selector := NewSelector(sourceIndex + "=" + source)
+	oldFlags, _, _ := s.GetAll(context.Background(), selector)
+
+	s.mx.Lock()
+	for key := range oldFlags {
+		if _, ok := flags[key]; !ok {
+			// flag has been deleted
+			s.logger.Debug(fmt.Sprintf("flag %s has been deleted from source %s", key, source))
```

Comment:
this could be at INFO level 🤔

Reply:
I'm not sure, I think this is perfectly normal behavior so I'm on the side of debug.